pypif_sdk/func/calculate_funcs.py

```diff
@@ -0,0 +1,258 @@
+import pypif
+import re
+from mendeleev import element
+from pypif.obj import *
+
+def calculate_ideal_atomic_percent(pif):
+    """
+    Calculates ideal atomic percents from a chemical formula string from a pif. Returns an appended pif with composition elements modified or added.
+    """
+    try:
+        pif.chemicalFormula
+    except NameError:
+        return pif
+    else:
+        expanded_formula = __expand_formula__(pif.chemicalFormula)
+        e_array = __create_emprical_compositional_array__(expanded_formula)
+        appended_e_array = __add_atomic_percents__(e_array)
+        for e in appended_e_array:
+            if __check_if_element_exists_in_pif_composition__(pif, e["symbol"]):
+                in_pif = __get_element_in_pif_composition__(pif, e["symbol"])
+                comp = in_pif[0]
+                pif.composition.pop(in_pif[1])
+                comp.idealAtomicPercent = e["atomic_percent"]
+                pif.composition.append(comp)
+            else:
+                comp = Composition()
+                comp.element = e["symbol"]
+                comp.idealAtomicPercent = e["atomic_percent"]
+                pif.composition.append(comp)
+        return pif
+
+def calculate_ideal_weight_percent(pif):
+    """
+    Calculates ideal atomic weight percents from a chemical formula string from a pif. Returns an appended pif with composition elements modified or added.
+    """
+    try:
+        pif.chemicalFormula
+    except NameError:
+        return pif
+    else:
+        expanded_formula = __expand_formula__(pif.chemicalFormula)
+        e_array = __create_emprical_compositional_array__(expanded_formula)
+        appended_e_array = __add_ideal_atomic_weights__(e_array)
+        a_array_with_pcts = __add_ideal_weight_percent__(appended_e_array)
+        for e in a_array_with_pcts:
+            if __check_if_element_exists_in_pif_composition__(pif, e["symbol"]):
+                in_pif = __get_element_in_pif_composition__(pif, e["symbol"])
+                comp = in_pif[0]
+                pif.composition.pop(in_pif[1])
+                comp.idealWeightPercent = e["weight_percent"]
+                pif.composition.append(comp)
+            else:
+                comp = Composition()
+                comp.element = e["symbol"]
+                comp.idealWeightPercent = e["weight_percent"]
+                pif.composition.append(comp)
+        return pif
+
+def __expand_formula__(formula_string):
```

Comment:
I _think_ its bad practice to use double leading and trailing underscores for user defined names, e.g. https://dbader.org/blog/meaning-of-underscores-in-python